src/utils/secp256k1.ts

```diff
@@ -0,0 +1,207 @@
+   * @returns uncompressed public key
+   */
+  public static async derivePublicKey(key: Uint8Array, relativePath: string): Promise<Uint8Array> {
+    const pathSegments = Secp256k1.parseAndValidateKeyDerivationPath(relativePath);
+
+    let currentPublicKey: Uint8Array;
+    if (key.length === 32) {
+      // private key is always 32 bytes
+      currentPublicKey = secp256k1.getPublicKey(key);
+    } else {
+      currentPublicKey = key;
+    }
+
+    for (const segment of pathSegments) {
+      const hash = await sha256.encode(Encoder.stringToBytes(segment));
+      currentPublicKey = Secp256k1.deriveChildPublicKey(currentPublicKey, hash);
+    }
+
+    return currentPublicKey;
+  }
+
+  /**
+   * Derives a hierarchical deterministic private key.
+   * @param relativePath `/` delimited path relative to the key given. e.g. 'a/b/c'
+   */
+  public static async derivePrivateKey(privateKey: Uint8Array, relativePath: string): Promise<Uint8Array> {
+    const pathSegments = Secp256k1.parseAndValidateKeyDerivationPath(relativePath);
+
+    let currentPrivateKey = privateKey;
+    for (const segment of pathSegments) {
+      const hash = await sha256.encode(Encoder.stringToBytes(segment));
+      currentPrivateKey = Secp256k1.deriveChildPrivateKey(currentPrivateKey, hash);
+    }
+
+    return currentPrivateKey;
+  }
+
+  /**
+   * Derives a child public key using the given tweak input.
+   */
+  public static deriveChildPublicKey(uncompressedPublicKey: Uint8Array, tweakInput: Uint8Array): Uint8Array {
+    const compressedPublicKey = false;
+    const derivedPublicKey = secp256k1Derivation.publicKeyTweakAdd(uncompressedPublicKey, tweakInput, compressedPublicKey);
+    return derivedPublicKey;
+  }
+
+  /**
+   * Derives a child private key using the given tweak input.
+   */
+  public static deriveChildPrivateKey(privateKey: Uint8Array, tweakInput: Uint8Array): Uint8Array {
+    // NOTE: passing in private key to v5.0.0 of `secp256k1.privateKeyTweakAdd()` has the side effect of morphing the input private key bytes
+    // before there is a fix for it (we can also investigate and submit a PR), we clone the private key to workaround
+    // `secp256k1.publicKeyTweakAdd()` does not have this side effect
+    const privateKeyClone = new Uint8Array(privateKey.length);
+    privateKeyClone.set(privateKey);
+
+    const derivedPrivateKey = secp256k1Derivation.privateKeyTweakAdd(privateKeyClone, tweakInput);
+    return derivedPrivateKey;
+  }
+
+  /**
+   * Parses the given key derivation path.
+   * @returns Path segments if successfully validate the derivation path.
+   * @throws {DwnError} with `DwnErrorCode.HdKeyDerivationPathInvalid` if derivation path fails validation.
+   */
+  private static parseAndValidateKeyDerivationPath(derivationPath: string): string[] {
+    const pathSegments = derivationPath.split('/');
+
+    if (pathSegments.length === 0 || pathSegments.includes('')) {
```

Comment:
Do we need to validate that path segments are in ASCII or anything like that?

Reply:
No, in fact, I have a subsequent PR that changed path to be an array of strings so that the path has less limitation (such as delimiter by `/`). I'll see if I can easily move that change into this PR. If not, you'll see it in the upcoming one.